File: csharp/Anatomy.cs
// Copyright (c) 2012 by Spirent Communications, Inc.
// All Rights Reserved
//
// By accessing or executing this software, you agree to be bound 
// by the terms of this agreement.
// 
// Redistribution and use of this software in source and binary forms,
// with or without modification, are permitted provided that the 
// following conditions are met:
//   1.  Redistribution of source code must contain the above copyright 
//       notice, this list of conditions, and the following disclaimer.
//   2.  Redistribution in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer
//       in the documentation and/or other materials provided with the
//       distribution.
//   3.  Neither the name Spirent Communications nor the names of its
//       contributors may be used to endorse or promote products derived
//       from this software without specific prior written permission.
//
// This software is provided by the copyright holders and contributors 
// [as is] and any express or implied warranties, including, but not 
// limited to, the implied warranties of merchantability and fitness for
// a particular purpose are disclaimed.  In no event shall Spirent
// Communications, Inc. or its contributors be liable for any direct, 
// indirect, incidental, special, exemplary, or consequential damages
// (including, but not limited to: procurement of substitute goods or
// services; loss of use, data, or profits; or business interruption)
// however caused and on any theory of liability, whether in contract, 
// strict liability, or tort (including negligence or otherwise) arising
// in any way out of the use of this software, even if advised of the
// possibility of such damage.

// File Name:                 Anatomy.cs
// Description:               This script demonstrates basic features 
//                            such as creating streams, generating traffic,
//                            enabling capture, saving realtime results
//                            to files, and retrieving results.

using StcCSharp;
using System;
using System.Collections.Generic;
using System.Threading;


public class Anatomy
{
    public void Run()
    {
        Stc.Init();       

        const bool ENABLE_CAPTURE = true;

        string StcVersion = Stc.Get("system1", "Version");
        Console.WriteLine("SpirentTestCenter system version:\t" + StcVersion);

        // Physical topology
        string szChassisIp1 = "10.29.0.49";
        string szChassisIp2 = "10.29.0.45";
        string txPortLoc = String.Format("//{0}/1/1", szChassisIp1);
        string rxPortLoc = String.Format("//{0}/1/1", szChassisIp2);

        // Create the root project object
        Console.WriteLine("Creating project ...");
        string hProject = Stc.Create("project", "system1");

        // Create ports
        Console.WriteLine("Creating ports ...");
        string hPortTx = Stc.Create("port", hProject);
        string hPortRx = Stc.Create("port", hProject);

        Dictionary<string, string> sm = new Dictionary<string, string>();
        sm.Clear();
        sm["location"] = txPortLoc;
        sm["useDefaultHost"] = "False";
        Stc.Config(hPortTx, sm);

        sm.Clear();
        sm["location"] = rxPortLoc;
        sm["useDefaultHost"] = "False";
        Stc.Config(hPortRx, sm);

        // Configure ethernet Fiber interface.
        Stc.Create("EthernetCopper", hPortTx);

        // Attach to ports.
        Console.WriteLine("Attaching to ports {0} and {1}", txPortLoc, rxPortLoc);
        Stc.Perform("AttachPorts");

        // Apply the configuration.
        Console.WriteLine("Apply configuration");
        Stc.Apply();

        // Initialize generator/analyzer.
        string hGenerator = Stc.Get(hPortTx, "children-Generator");
        string generatorState = Stc.Get(hGenerator, "state");
        Console.WriteLine("Stopping Generator -current state " + generatorState);
        sm.Clear();
        sm["generatorList"] = hGenerator;
        Stc.Perform("generatorStop", sm);

        string hAnalyzer = Stc.Get(hPortRx, "children-Analyzer");
        string analyzerState = Stc.Get(hAnalyzer, "state");
        Console.WriteLine("Stopping Analyzer -current state " + analyzerState);
        sm.Clear();
        sm["analyzerList"] = hAnalyzer;
        Stc.Perform("analyzerStop", sm);

        // Create a stream block. FrameConfig with blank double quotes Clears the frame out.
        Console.WriteLine("Configuring stream block ...");
        string hStreamBlock = Stc.Create("streamBlock", hPortTx);
        sm.Clear();
        sm["insertSig"] = "true";
        sm["frameConfig"] = "";
        sm["frameLengthMode"] = "FIXED";
        sm["maxFrameLength"] = "1200";
        sm["FixedFrameLength"] = "256";
        Stc.Config(hStreamBlock, sm);

        // Add an EthernetII Protocol Data Unit (PDU).
        Console.WriteLine("Adding headers");
        string ethPDU = Stc.Create("ethernet:EthernetII", hStreamBlock);
        sm.Clear();
        sm["name"] = "sb1_eth";
        sm["srcMac"] = "00:00:20:00:00:00";
        sm["dstMac"] = "00:00:00:00:00:00";
        Stc.Config(ethPDU, sm);

        // Use modifier to generate multiple streams.
        Console.WriteLine("Creating Modifier on Stream Block ...");
        string hRangeModifier = Stc.Create("RangeModifier", hStreamBlock);
        sm.Clear();
        sm["ModifierMode"] = "INCR";
        sm["Mask"] = "0000FFFFFFFF";
        sm["StepValue"] = "000000000001";
        sm["Data"] = "000000000000";
        sm["RecycleCount"] = "4294967295";
        sm["RepeatCount"] = "0";
        sm["DataType"] = "BYTE";
        sm["EnableStream"] = "FALSE";
        sm["Offset"] = "0";
        sm["OffsetReference"] = "sb1_eth.dstMac";
        Stc.Config(hRangeModifier, sm);

        // Configure generator
        Console.WriteLine("Configuring Generator");
        string hGeneratorConfig = Stc.Get(hGenerator, "children-GeneratorConfig");

        sm.Clear();
        sm["DurationMode"] = "SECONDS";
        sm["BurstSize"] = "1";
        sm["Duration"] = "100";
        sm["LoadMode"] = "FIXED";
        sm["FixedLoad"] = "25";
        sm["LoadUnit"] = "PERCENT_LINE_RATE";
        sm["SchedulingMode"] = "PORT_BASED";
        Stc.Config(hGeneratorConfig, sm);

        // Analyzer Configuration
        Console.WriteLine("Configuring Analyzer");
        string hAnalyzerConfig = Stc.Get(hAnalyzer, "children-AnalyzerConfig");

        // Subscribe to realtime results
        Console.WriteLine("Subscribe to results");
        sm.Clear();
        sm["Parent"] = hProject;
        sm["ConfigType"] = "Analyzer";
        sm["resulttype"] = "AnalyzerPortResults";
        sm["filenameprefix"] = "Analyzer_Port_Results";
        Stc.Subscribe(sm);

        sm.Clear();
        sm["Parent"] = hProject;
        sm["ConfigType"] = "Generator";
        sm["resulttype"] = "Generator";
        sm["filenameprefix"] = "Generator_Port_Counter";
        Stc.Subscribe(sm);

        // Configure Capture.
        string hCapture = "";
        if (ENABLE_CAPTURE)
        {
            Console.WriteLine("\nStarting Capture...");

            // Create a capture object. Automatically created.
            hCapture = Stc.Get(hPortRx, "children-capture");
            sm.Clear();
            sm["mode"] = "REGULAR_MODE";
            sm["srcMode"] = "TX_RX_MODE";
            Stc.Config(hCapture, sm);
            sm.Clear();
            sm["captureProxyId"] = hCapture;
            Stc.Perform("CaptureStart", sm);
        }

        // Apply configuration.  
        Console.WriteLine("Apply configuration");
        Stc.Apply();

        // Save the configuration as an XML file for later import into the GUI.
        Console.WriteLine("\nSave configuration as an XML file.");
        Stc.Perform("SaveAsXml");

        // Start the analyzer and generator.
        Console.WriteLine("Start Analyzer");
        sm.Clear();
        sm["AnalyzerList"] = hAnalyzer;
        Stc.Perform("AnalyzerStart", sm);
        analyzerState = Stc.Get(hAnalyzer, "state");
        Console.WriteLine("Current analyzer state " + analyzerState);

        Thread.Sleep(2000);

        Console.WriteLine("Start Generator");
        sm.Clear();
        sm["GeneratorList"] = hGenerator;
        Stc.Perform("GeneratorStart", sm);
        generatorState = Stc.Get(hGenerator, "state");
        Console.WriteLine("Current generator state " + generatorState);

        Console.WriteLine("Wait 5 seconds ...");
        Thread.Sleep(5000);

        analyzerState = Stc.Get(hAnalyzer, "state");
        generatorState = Stc.Get(hGenerator, "state");
        Console.WriteLine("Current analyzer state " + analyzerState);
        Console.WriteLine("Current generator state " + generatorState);

        // Stop the analyzer.  
        Console.WriteLine("Stop Analyzer");
        sm.Clear();
        sm["AnalyzerList"] = hAnalyzer;
        Stc.Perform("AnalyzerStop", sm);
        Thread.Sleep(1000);

        // Display some statistics.

        Console.WriteLine("Frames Counts:");
        // Example of Direct-Descendant Notation ( DDN ) syntax. ( DDN starts with an object reference )  
        string sigFrameCount = Stc.Get(String.Format("{0}.AnalyzerPortResults(1)", hAnalyzer), "sigFrameCount");
        string totalFrameCount = Stc.Get(String.Format("{0}.AnalyzerPortResults(1)", hAnalyzer), "totalFrameCount");        
        Console.WriteLine("\tSignature frames: {0}", sigFrameCount);
        Console.WriteLine("\tTotal frames: {0}", totalFrameCount);

        // Example of Descendant-Attribute Notation ( DAN ) syntax. ( using explicit indeces )
        string minFrameLength = Stc.Get(hPortRx, "Analyzer(1).AnalyzerPortResults(1).minFrameLength");
        Console.WriteLine("\tMinFrameLength: {0}", minFrameLength);
        // Notice indexing is not necessary since there is only 1 child. 
        string maxFrameLength = Stc.Get(hPortRx, "Analyzer.AnalyzerPortResults.maxFrameLength");
        Console.WriteLine("\tMaxFrameLength: {0}", maxFrameLength);

        if (ENABLE_CAPTURE)
        {
            Console.WriteLine("Retrieving Captured frames...");
            sm.Clear();
            sm["captureProxyId"] = hCapture;
            Stc.Perform("CaptureStop", sm);

            // Save captured frames to a file.
            sm.Clear();
            sm["captureProxyId"] = hCapture;
            sm["FileName"] = "capture.pcap";
            sm["FileNameFormat"] = "PCAP";
            sm["IsScap"] = "FALSE";
            Stc.Perform("CaptureDataSave", sm);

            string capPackets = Stc.Get(hCapture, "PktCount");
            Console.WriteLine("Captured frames:\t" + capPackets);
        }

        // Disconnect from chassis, release ports, and reset configuration.
        Console.WriteLine("Release ports and disconnect from chassis");
        Stc.Perform("ChassisDisconnectAll");
        Stc.Perform("ResetConfig");

        // Delete configuration
        Console.WriteLine("Deleting project");
        Stc.Delete(hProject);
    }

    static void Main(string[] args)
    {
        try
        {            
            new Anatomy().Run();
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.Message);            
        }
    }
}

File: java/Anatomy.java
// Copyright (c) 2010 by Spirent Communications, Inc.
// All Rights Reserved
//
// By accessing or executing this software, you agree to be bound 
// by the terms of this agreement.
// 
// Redistribution and use of this software in source and binary forms,
// with or without modification, are permitted provided that the 
// following conditions are met:
//   1.  Redistribution of source code must contain the above copyright 
//       notice, this list of conditions, and the following disclaimer.
//   2.  Redistribution in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer
//       in the documentation and/or other materials provided with the
//       distribution.
//   3.  Neither the name Spirent Communications nor the names of its
//       contributors may be used to endorse or promote products derived
//       from this software without specific prior written permission.
//
// This software is provided by the copyright holders and contributors 
// [as is] and any express or implied warranties, including, but not 
// limited to, the implied warranties of merchantability and fitness for
// a particular purpose are disclaimed.  In no event shall Spirent
// Communications, Inc. or its contributors be liable for any direct, 
// indirect, incidental, special, exemplary, or consequential damages
// (including, but not limited to: procurement of substitute goods or
// services; loss of use, data, or profits; or business interruption)
// however caused and on any theory of liability, whether in contract, 
// strict liability, or tort (including negligence or otherwise) arising
// in any way out of the use of this software, even if advised of the
// possibility of such damage.

// File Name:                 Anatomy.java
// Description:               This script demonstrates basic features 
//                            such as creating streams, generating traffic,
//                            enabling capture, saving realtime results
//                            to files, and retrieving results.

import java.util.HashMap;
import java.util.Map;

import com.spirent.stc;

public class Anatomy {
	
	public static void main(String[] args)
	{
		try {

			stc.Init();
		
			boolean ENABLE_CAPTURE = true;
			
			String stcVersion = stc.Get("system1", "Version");
			System.out.println("SpirentTestCenter system version:\t" + stcVersion);

			// Physical topology
			String szChassisIp1 = "10.29.0.49";
			String szChassisIp2 = "10.29.0.45";
			String txPortLoc = String.format("//%s/%s/%s", szChassisIp1, 1, 1);
			String rxPortLoc = String.format("//%s/%s/%s", szChassisIp2, 1, 1);
		
			// Create the root project object
			System.out.println("Creating project ...");
			String hProject = stc.Create("project", "system1");

			// Create ports
			System.out.println("Creating ports ...");
			String hPortTx = stc.Create("port", hProject);
			String hPortRx = stc.Create("port", hProject);

			Map<String, String> sm = new HashMap<String, String>();
			sm.clear();
			sm.put("location", txPortLoc);
			sm.put("useDefaultHost", "False");
			stc.Config(hPortTx, sm);

			sm.clear();
			sm.put("location", rxPortLoc);
			sm.put("useDefaultHost", "False");
			stc.Config(hPortRx, sm);

			// Configure ethernet Fiber interface.
			stc.Create("EthernetCopper", hPortTx);

			// Attach ports. 
			// Connects to chassis, reserves ports and sets up port mappings all in one step.
			// By default, connects to all previously created ports.			
			System.out.println(String.format("Attaching ports %s %s", txPortLoc, rxPortLoc));
			stc.Perform("AttachPorts");
			
			// Apply the configuration.
			System.out.println("Apply configuration");
			stc.Apply();

			// Initialize generator/analyzer.
			String hGenerator = stc.Get(hPortTx, "children-Generator");
			String generatorState = stc.Get(hGenerator, "state");
			System.out.println("Stopping Generator -current state " + generatorState);
			sm.clear();
			sm.put("generatorList", hGenerator);
			stc.Perform("generatorStop", sm);

			String hAnalyzer = stc.Get(hPortRx, "children-Analyzer");
			String analyzerState = stc.Get(hAnalyzer, "state");
			System.out.println("Stopping Analyzer -current state " + analyzerState);
			sm.clear();
			sm.put("analyzerList", hAnalyzer);
			stc.Perform("analyzerStop", sm);

			// Create a stream block. FrameConfig with blank double quotes clears the frame out.
			System.out.println("Configuring stream block ...");
			String hStreamBlock = stc.Create("streamBlock", hPortTx);
			sm.clear();
			sm.put("insertSig", "true");
			sm.put("frameConfig", "");
			sm.put("frameLengthMode", "FIXED");
			sm.put("maxFrameLength", "1200");
			sm.put("FixedFrameLength", "256");
			stc.Config(hStreamBlock, sm);

			// Add an EthernetII Protocol Data Unit (PDU).
			System.out.println("Adding headers");
			String ethPDU = stc.Create("ethernet:EthernetII", hStreamBlock);
			sm.clear();
			sm.put("name", "sb1_eth");
			sm.put("srcMac", "00:00:20:00:00:00");
			sm.put("dstMac", "00:00:00:00:00:00");
			stc.Config(ethPDU, sm);

			// Use modifier to generate multiple streams.
			System.out.println("Creating Modifier on Stream Block ...");
			String hRangeModifier = stc.Create("RangeModifier", hStreamBlock);
			sm.clear();
			sm.put("ModifierMode", "INCR");
			sm.put("Mask", "0000FFFFFFFF");
			sm.put("StepValue", "000000000001");
			sm.put("Data", "000000000000");
			sm.put("RecycleCount", "4294967295");
			sm.put("RepeatCount", "0");
			sm.put("DataType", "BYTE");
			sm.put("EnableStream", "FALSE");
			sm.put("Offset", "0");
			sm.put("OffsetReference", "sb1_eth.dstMac");
			stc.Config(hRangeModifier, sm);
			
			// Configure generator
			System.out.println("Configuring Generator");
			String hGeneratorConfig = stc.Get(hGenerator, "children-GeneratorConfig");

			sm.clear();
			sm.put("DurationMode", "SECONDS");
			sm.put("BurstSize", "1");
			sm.put("Duration", "100");
			sm.put("LoadMode", "FIXED");
			sm.put("FixedLoad", "25");
			sm.put("LoadUnit", "PERCENT_LINE_RATE");
			sm.put("SchedulingMode", "PORT_BASED");
			stc.Config(hGeneratorConfig, sm);

			// Analyzer Configuration
			System.out.println("Configuring Analyzer");
			String hAnalyzerConfig = stc.Get(hAnalyzer, "children-AnalyzerConfig");

			// Subscribe to realtime results
			System.out.println("Subscribe to results");
			sm.clear();
			sm.put("Parent", hProject);
			sm.put("ConfigType", "Analyzer");
			sm.put("resulttype", "AnalyzerPortResults");
			sm.put("filenameprefix", "Analyzer_Port_Results");
			stc.Subscribe(sm);

			sm.clear();
			sm.put("Parent", hProject);
			sm.put("ConfigType", "Generator");
			sm.put("resulttype", "Generator");
			sm.put("filenameprefix", "Generator_Port_Counter");
			stc.Subscribe(sm);

			// Configure Capture.
			String hCapture = "";
			if (ENABLE_CAPTURE)
			{
				System.out.println("\nStarting Capture...");
  
				// Create a capture object. Automatically created.
				hCapture = stc.Get(hPortRx, "children-capture");
				sm.clear();
				sm.put("mode", "REGULAR_MODE");
				sm.put("srcMode", "TX_RX_MODE");
				stc.Config(hCapture, sm);
				sm.clear();
				sm.put("captureProxyId", hCapture);
				stc.Perform("CaptureStart", sm);
			}

			// Apply configuration.  
			System.out.println("Apply configuration");
			stc.Apply();

			// Save the configuration as an XML file for later import into the GUI.
			System.out.println("\nSave configuration as an XML file.");
			stc.Perform("SaveAsXml");

			// Start the analyzer and generator.
			System.out.println("Start Analyzer");
			sm.clear();
			sm.put("AnalyzerList", hAnalyzer);
			stc.Perform("AnalyzerStart", sm);
			analyzerState = stc.Get(hAnalyzer, "state");
			System.out.println("Current analyzer state " + analyzerState);

			Thread.sleep(2000);

			System.out.println("Start Generator");
			sm.clear();
			sm.put("GeneratorList", hGenerator);
			stc.Perform("GeneratorStart", sm);
			generatorState = stc.Get(hGenerator, "state");
			System.out.println("Current generator state " + generatorState);

			System.out.println("Wait 5 seconds ...");
			Thread.sleep(5000);

			analyzerState = stc.Get(hAnalyzer, "state");
			generatorState = stc.Get(hGenerator, "state");
			System.out.println("Current analyzer state " + analyzerState);
			System.out.println("Current generator state " + generatorState);
			
			// Stop the analyzer.  
			System.out.println("Stop Analyzer");
			sm.clear();
			sm.put("AnalyzerList", hAnalyzer);
			stc.Perform("AnalyzerStop", sm);
			Thread.sleep(1000);

			// Display some statistics.

			System.out.println("Frames Counts:");
			// Example of Direct-Descendant Notation ( DDN ) syntax. ( DDN starts with an object reference )  
			String sigFrameCount = stc.Get(String.format("%s.AnalyzerPortResults(1)", hAnalyzer), "sigFrameCount");
			String totalFrameCount = stc.Get(String.format("%s.AnalyzerPortResults(1)", hAnalyzer), "totalFrameCount");
			System.out.println("\tSignature frames: " + sigFrameCount);
			System.out.println("\tTotal frames: " + totalFrameCount);

			// Example of Descendant-Attribute Notation ( DAN ) syntax. ( using explicit indeces )
			String minFrameLength = stc.Get(hPortRx, "Analyzer(1).AnalyzerPortResults(1).minFrameLength");
			System.out.println("\tMinFrameLength: " + minFrameLength);
			// Notice indexing is not necessary since there is only 1 child. 
			String maxFrameLength = stc.Get(hPortRx, "Analyzer.AnalyzerPortResults.maxFrameLength");
			System.out.println("\tMaxFrameLength: " + maxFrameLength);
			
			if (ENABLE_CAPTURE)
			{
				System.out.println("Retrieving Captured frames...");
				sm.clear();
				sm.put("captureProxyId", hCapture);
				stc.Perform("CaptureStop", sm);
    
				// Save captured frames to a file.
				sm.clear();
				sm.put("captureProxyId", hCapture);
				sm.put("FileName", "capture.pcap");
				sm.put("FileNameFormat", "PCAP");
				sm.put("IsScap", "FALSE");
				stc.Perform("CaptureDataSave", sm);
				
				String capPackets = stc.Get(hCapture, "PktCount");
				System.out.println("Captured frames:\t" + capPackets);
			}

			// Disconnect from chassis, release ports, and reset configuration.
			System.out.println("Release ports and disconnect from chassis");		
			stc.Perform("ChassisDisconnectAll");			
			stc.Perform("ResetConfig");
			
			// Delete configuration
			System.out.println("Deleting project");
			stc.Delete(hProject);

		} catch (Exception catchEx) {
			System.out.println("Exception: " + catchEx);
		}
	}
}
